parlai/utils/fp16.py

```diff
@@ -0,0 +1,677 @@
+            targets,
+            ignore_index=self.NULL_IDX,
+            reduction=self.reduction,
+        )
+
+
+def clip_grad_norm(params, max_norm):
+    """
+    Clips grad norm.
+    """
+    params = list(params)
+    if len(params) == 1:
+        p = params[0]
+        grad_norm = torch.norm(p)
+        if grad_norm > max_norm > 0:
+            clip_coef = max_norm / (grad_norm + 1e-6)
+            p.mul_(clip_coef)
+        return grad_norm
+    elif max_norm > 0:
+        return torch.nn.utils.clip_grad_norm_(params, max_norm)
+    else:
+        return torch.sqrt(
+            sum(p.grad.data.norm() ** 2 for p in params if p.grad is not None)
+        )
+
+
+def has_overflow(grad_norm):
+    """
+    Detect inf and NaN in grad_norm.
+    """
+    if grad_norm == float('inf') or grad_norm != grad_norm:
+        return True
+    return False
+
+
+###################################################
+## APEX Wrapper
+###################################################
+
+
+def fp16_optimizer_wrapper(
+    optimizer: torch.optim.Optimizer,  # type: ignore
+    verbose: bool = False,
+    dynamic_loss_scale: bool = True,
+    loss_initial_scale: float = 2.0 ** 17,
+):
+    """
+    Wrap the an optimizer with FP16 loss scaling protection.
+
+    Requires apex to be installed. Will throw an ImportError if it is not.
+
+    :param optimizer:
+        Any torch optimizer
+    :param bool verbose:
+        Enables verbose output in the FP16 optimizer. Turning this on can help
+        debug when FP16 is underperforming.
+    :param bool dynamic_loss_scaling:
+        FP16 requires loss scaling to avoid underflows. It is recommended this
+        stays on, but advanced users may want it off.
+    :param float loss_initial_scale:
+        Initial loss scaling. Default chosen empirically, but models with very low
+        or high loss values may need this adjusted. Stick with powers of 2.
+
+    :returns:
+        An APEX FP16 optimizer. Please note this has different requirements on
+        how backward() and step() are called.
+    """
+    try:
+        import apex.fp16_utils
+    except ImportError:
+        raise ImportError(
+            'No fp16 support without apex. Please install it from '
+            'https://github.com/NVIDIA/apex'
+        )
+    return apex.fp16_utils.FP16_Optimizer(
+        optimizer,
+        dynamic_loss_scale=dynamic_loss_scale,
+        verbose=verbose,
+        # TODO: We may later want to remove this flag. Right now it
+        # empirically improves the first few backward passes, but future APEX
+        # improvements may make this unnecessary.
+        dynamic_loss_args={'init_scale': loss_initial_scale},
+    )
+
+
+def fp16_apex_available() -> bool:
+    try:
+        import apex.fp16_utils  # noqa: F401
+
+        return True
+    except ImportError:
+        warn_once(
+            'You set --fp16 true, but fp16 is unavailable. To use fp16, please '
```

Comment:
should we modify this warning to indicate that one could also set `--fp16-impl` to `mem_efficient`?